src/aerovaldb/utils/uri.py

```diff
@@ -7,38 +7,55 @@ def extract_substitutions(template: str):
     """
     For a python template string, extracts the names between curly brackets:
 
-    For example 'blah blah {test} blah {test2}' returns [test, test2]
+    For example 'blah blah {test} blah {test2}' returns ["test", "test2"]
     """
     return re.findall(r"\{([a-zA-Z-]*?)\}", template)
 
 
-def parse_formatted_string(template: str, s: str) -> dict:
-    """Match s against a python format string, extracting the
-    parameter values from the format string in a dictionary.
-
-    Note
-    ----
-    Only supports {named_parameter} style format.
-    """
-
-    # First split on any keyword arguments, note that the names of keyword arguments will be in the
-    # 1st, 3rd, ... positions in this list
-    tokens = re.split(r"\{([a-zA-Z-]*?)\}", template)
-    # keywords = tokens[1::2]
+def parse_formatted_string(template: str, s: str):
     keywords = extract_substitutions(template)
-    # Now replace keyword arguments with named groups matching them. We also escape between keyword
-    # arguments so we support meta-characters there. Re-join tokens to form our regexp pattern
 
-    tokens[1::2] = map("(?P<{}>[^/]*)".format, keywords)
-    tokens[0::2] = map(re.escape, tokens[0::2])
-    pattern = "".join(tokens)
+    pattern = "(" + "|".join([re.escape("{" + k + "}") for k in keywords]) + ")"
+    segments = re.split(pattern, template)
 
-    # Use our pattern to match the given string, raise if it doesn't match
-    if not (match := re.match(pattern, s)):
-        raise Exception("Format string did not match")
-
-    # Return a dict with all of our keywords and their values
-    return {x: match.group(x) for x in keywords}
+    result = {}
+    while len(segments) > 0:
+        token = segments[0]
+        if token.startswith("{"):
+            if s[0] == '"':
+                extr = s.split('"')[1]
+                s = s[(len(extr) + 2) :]
+            else:
+                if (len(segments) >= 2 and segments[1] != "") or (len(segments) > 2):
+                    ls: list[str] = []
+                    esc_flag = False
+                    while True:
+                        char = s[len(ls)]
+                        if esc_flag == True:
+                            ls.append(char)
+                            esc_flag = False
+                            continue
+                        if char == "\\":
+                            esc_flag = True
+                            continue
+
+                        ls.append(char)
+                        if s[len(ls) :].startswith(segments[1]):
+                            break
+
+                    extr = "".join(ls)
+                    s = s[(len(extr)) :]
+                else:
+                    extr = s
+
+            result[token.replace("{", "").replace("}", "")] = extr
+        else:
+            if not s.startswith(token):
+                raise Exception("Format string did not match")
+            s = s[len(token) :]
+
+        segments = segments[1:]
+    return result
```

Comment:
This looks like voodoo! Please add lots of comments what the different parts intend to do.